### PR TITLE
Set process title when running cron job

### LINKED ISSUE #34321: [Issue] Set process title when running cron job
This issue is automatically created based on existing pull request: magento/magento2#34186: Set process title when running cron job

---------
### Description
When managing a busy server, it's helpful to understand what each process is doing. Currently there is no visibility for a systems administrator to see what task the `bin/magento cron:run` process is currently performing. The most information that's available is that some job (or set of jobs) is running within a PHP process that was started via the system task scheduler (read: cron daemon). Because of the way that Magento spawns additional processes for some groups (depending on their configuration) it is possible to sometimes narrow down the list of possible tasks because the group name is listed in the command arguments.

This pull request sets the title of the currently-running process to clearly list the group and job names.

<details><summary>Before screen-shot</summary>

Screenshot_2021-10-04_22-59-50

</details><details><summary>After screen-shot</summary>

Screenshot_2021-10-04_22-56-17

</details>

### Related Pull Requests
*None*

### Manual testing scenarios
1. Run the standard Magento cron with `bin/magento cron:run` when several long-running tasks are scheduled
1. Observe lack of (or presence of after this change) information about which job is running in the process tree (using tools like `ps` or `htop` on Unix - untested on Windows, but the functionality should work there too)

### Questions or comments
Should the path to Magento be listed in the process title too? If there are multiple Magento instances on a single server, they should be running under different users. I can't think of a sensible non-development environment where multiple Magento instances run under the same system user.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by fredden**
### Description
When managing a busy server, it's helpful to understand what each process is doing. Currently there is no visibility for a systems administrator to see what task the `bin/magento cron:run` process is currently performing. The most information that's available is that some job (or set of jobs) is running within a PHP process that was started via the system task scheduler (read: cron daemon). Because of the way that Magento spawns additional processes for some groups (depending on their configuration) it is possible to sometimes narrow down the list of possible tasks because the group name is listed in the command arguments.

This pull request sets the title of the currently-running process to clearly list the group and job names.

<details><summary>Before screen-shot</summary>

Screenshot_2021-10-04_22-59-50

</details><details><summary>After screen-shot</summary>

Screenshot_2021-10-04_22-56-17

</details>

### Related Pull Requests
*None*

### Manual testing scenarios
1. Run the standard Magento cron with `bin/magento cron:run` when several long-running tasks are scheduled
1. Observe lack of (or presence of after this change) information about which job is running in the process tree (using tools like `ps` or `htop` on Unix - untested on Windows, but the functionality should work there too)

### Questions or comments
Should the path to Magento be listed in the process title too? If there are multiple Magento instances on a single server, they should be running under different users. I can't think of a sensible non-development environment where multiple Magento instances run under the same system user.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#34321: Set process title when running cron job

### CODE REVIEW

**Comment by fredden** [comment]
@magento run all tests

**Comment by hostep** [comment]
@fredden: very nice, this sounds super useful!

Can you put a screenshot here, of how it looks with `ps` or `top`/`htop` ?

> Should the path to Magento be listed in the process title too? If there are multiple Magento instances on a single server, they should be running under different users. I can't think of a sensible non-development environment where multiple Magento instances run under the same system user.

We do occasionally have this situation, it's usually for some demo shops or something like that, never for a real production shop, so there are still valid cases I believe.

**Comment by fredden** [comment]
Thanks for your input @hostep. I've added before & after pictures to the main pull request description. I'm still somewhat undecided about what the process title should contain; I'm confident that including the process group & current job name will be helpful. I don't know if there's enough room for a full path too. Perhaps the path should go at the end?

**Comment by hostep** [comment]
Looks really useful, but it might introduce backwards incompatibilities, for example, [the following code](https://github.com/magento/ece-tools/blob/03a900824e1600a40dbfef1d4128840b91929a87/src/Util/BackgroundProcess.php#L49-L56) will probably stop working ...

If you could somehow try to append the extra info to the original output, it might be better?
For example:
```
/usr/local/bin/php /var/www/magento2/bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1 (group: index, job: indexer_reindex_all_invalid)
```

But not sure how complicated this would be, not sure if you can fetch the original output using `cli_get_process_title` to be able to append the extra info?

**Comment by fredden** [comment]
Thanks for your feedback, Pieter. Based on this I've updated the pull request to preserve the original command line arguments and add the new information at the end. I've updated the screen-shots in the main description.

I'm open to including `PHP_BINARY` at the start of the new title; let me know if that'd be useful too.

**Comment by fredden** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento create issue

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php [hunk 1]
```diff
@@ -128,6 +129,16 @@ class ProcessCronQueueObserver implements ObserverInterface
      */
     protected $dateTime;
 
+    /**
+     * @var Environment
+     */
+    protected Environment $environment;
```
Comment (sidolov) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: PLease, make property private

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php [hunk 2]
```diff
@@ -128,6 +129,16 @@ class ProcessCronQueueObserver implements ObserverInterface
...
+
+    /**
+     * @var string
+     */
+    protected string $originalProcessTitle;
```
Comment (sidolov) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: PLease, make property private

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE, Functional Tests B2B, Functional Tests CE

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Magento Health Index, Database Compare, Functional Tests CE, Functional Tests B2B, Functional Tests EE, Integration Tests

**Comment by Den4ik** [comment]
Hi @fredden. Is it possible don't use global variables at `Test` file?

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B,Functional Tests EE

**Comment by engcom-Charlie** [comment]
Hi @fredden,

Thank you for your contribution!

The test cases which are failing now are not part of this PR or not failing because of the PR changes. 

It seems, the requested changes have been taken care so requesting for the review hence moving the PR in Pending Review.

Thank you!

**Comment by fredden** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
It seems that bot moved this PR from Pending Review to Changes Requested hence moving it back for review.

Thank you!

**Comment by Den4ik** [comment]
@magento run Functional Tests EE, Functional Tests B2B, Functional Tests CE

**Review by sidolov — APPROVED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run Functional Tests CE

**Comment by engcom-Echo** [comment]
As on last two Functional Tests CE  run `MC-14271: Admin user is not able to delete the own account` test is failing hence moving to `Extended Testing` to check the build failures.

**Comment by engcom-Echo** [comment]
Functional Tests CE failure are different on last two run.
Run1
image

Run2
image

Hence moving to `Merge In Progress`.

**Pull request merged**